homeassistant/components/simplefin/config_flow.py

```diff
@@ -0,0 +1,110 @@
+"""Config flow for SimpleFIN integration."""
+
+from collections.abc import Mapping
+from typing import Any
+
+from simplefin4py import SimpleFin
+from simplefin4py.exceptions import (
+    SimpleFinAuthError,
+    SimpleFinClaimError,
+    SimpleFinInvalidAccountURLError,
+    SimpleFinInvalidClaimTokenError,
+    SimpleFinPaymentRequiredError,
+)
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.config_entries import ConfigFlowResult
+from homeassistant.const import CONF_API_TOKEN
+
+from .const import CONF_ACCESS_URL, DOMAIN, LOGGER
+
+
+async def _validate_or_obtain_access_url(input_string: str) -> str:
+    """Validate the input string as an access URL or a claim token and fetch data using SimpleFin.
+
+    Args:
+        input_string (str): The input_string will either be a URL or a base64 encoded claim_token
+
+    Returns:
+        str: The validated access URL - (or throws an error)
+
+    Raises:
+        SimpleFinInvalidAccountURLError: If the input string is an invalid access URL.
+        SimpleFinPaymentRequiredError
+        SimpleFinAuthError
+        SimpleFinInvalidClaimTokenError: If the input string is an invalid claim token.
+        SimpleFinClaimError: If there's an error in claim token processing.
+
+    """
+
+    if not input_string.startswith("http"):
+        LOGGER.info("[Setup Token] - Claiming Access URL")
+        access_url = await SimpleFin.claim_setup_token(input_string)
+
+    else:
+        LOGGER.info("[Access Url] - 'http' string detected")
+        access_url = input_string
+        LOGGER.info("[Access Url] - validating access url")
+        SimpleFin.decode_access_url(access_url)
+
+    LOGGER.info("[Access Url] - Fetching data")
+    simple_fin = SimpleFin(access_url=access_url)
+    await simple_fin.fetch_data()
+    return access_url
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for the initial setup of a SimpleFIN integration."""
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Prompt user for SimpleFIN API credentials."""
+        errors = {}
+
+        if user_input is not None:
+            try:
+                user_input[CONF_API_TOKEN] = await _validate_or_obtain_access_url(
+                    user_input[CONF_API_TOKEN]
+                )
+            except SimpleFinInvalidAccountURLError:
+                errors["base"] = "url_error"
+            except SimpleFinInvalidClaimTokenError:
+                errors["base"] = "invalid_claim_token"
+            except SimpleFinClaimError:
+                errors["base"] = "claim_error"
+            except SimpleFinPaymentRequiredError:
+                errors["base"] = "payment_required"
+            except SimpleFinAuthError:
+                errors["base"] = "auth_error"
+
+            entry = await self.async_set_unique_id(user_input[CONF_API_TOKEN])
+            if entry:
```

Comment:
This is not the correct way to check if we're doing reauth or not, instead do like this:
https://github.com/home-assistant/core/blob/a1332c2240af29ae8e0d04fab7b2c868c48cad3f/homeassistant/components/airvisual_pro/config_flow.py#L89-L95